registry/src/lib.rs

```diff
@@ -0,0 +1,422 @@
+pub struct AtomicUpdateParam {
+    /// The new module reference.
+    pub module: ModuleReference,
+    /// Optional entry point to call in the new module after the upgrade.
+    pub migrate: Option<(OwnedEntrypointName, OwnedParameter)>,
+    /// The contract address to natively upgrade.
+    pub contract_address: ContractAddress,
+}
+
+/// The parameter type for the contract function `upgradeNatively`.
+#[derive(Debug, Serialize, SchemaType)]
+pub struct UpgradeParams {
+    /// The new module reference.
+    pub module: ModuleReference,
+    /// Optional entry point to call in the new module after the upgrade.
+    pub migrate: Option<(OwnedEntrypointName, OwnedParameter)>,
+}
+
+/// This entry point ensures, that the old and the new contracts are aware of their states in the registry by calling the `upgradeNatively` and the `unregister` hooks.
+/// ATTENTION: If you want to upgrade the `UmbrellaFeeds` contract, use this function to natively upgrade the `UmbrellaFeeds` contract.
+/// Using the native upgradability mechanism for the `UmbrellaFeeds` contract is necessary to not break the `UmbrellaFeedsReader` contracts which include references to the `UmbrellaFeeds` contract.
+#[receive(
+    contract = "registry",
+    name = "atomicUpdate",
+    parameter = "AtomicUpdateParam",
+    error = "CustomContractError",
+    enable_logger,
+    mutable
+)]
+fn atomic_update<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<State<S>, StateApiType = S>,
+    logger: &mut impl HasLogger,
+) -> Result<(), CustomContractError> {
+    ensure_eq!(
+        ctx.sender(),
+        host.state().owner.ok_or(CustomContractError::NoOwner)?,
+        CustomContractError::UnauthorizedAccount
+    );
+
+    let params: AtomicUpdateParam = ctx.parameter_cursor().get()?;
+
+    let upgrade_params = UpgradeParams {
+        module: params.module,
+        migrate: params.migrate,
+    };
+
+    // Execute the `upgradeNatively()` hook
+    host.invoke_contract::<UpgradeParams>(
+        &params.contract_address,
+        &upgrade_params,
+        EntrypointName::new_unchecked("upgradeNatively"),
+        Amount::zero(),
+    )?;
+
+    let name = host.invoke_contract_read_only(
+        &params.contract_address,
+        &Parameter::empty(),
+        EntrypointName::new_unchecked("getName"),
+        Amount::zero(),
+    )?;
+
+    let name: String = name
+        .ok_or(CustomContractError::InvokeContractError)?
+        .get()?;
+
+    let old_contract = host
+        .state_mut()
+        .registry
+        .insert(name.clone(), params.contract_address);
+
+    // Only if another `old_contract` was already registered, execute the `unregister` hook
+    if let Some(old_contract) = old_contract {
+        // Execute the `unRegister()` hook
+        host.invoke_contract(
+            &old_contract,
+            &Parameter::empty(),
+            EntrypointName::new_unchecked("unregister"),
+            Amount::zero(),
+        )?;
```

Comment:
I guess it's deliberate to fail if unregister fails?

Can you explain why we want this?

Reply:
The `unregister` hook could have been used to e.g. `self_destruct` the old contract on Ethereum which would resulted in some gas refund.

If this hook fails then the old contract wasn't deployed with this `unregister` hook (most likely) and as a result, it wasn't the intention to use `atomicUpdate`. Failing this call makes it more obvious if something wasn't intentioned to be used in a particular case.